handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java

```diff
@@ -0,0 +1,203 @@
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        init(ctx);
+    }
+
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        destroy();
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        destroy();
+        super.channelInactive(ctx);
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        readFinishTime = System.nanoTime();
+        if (timeoutShouldBeCancelled(ctx, msg)) {
+            disable();
+        }
+        super.channelRead(ctx, msg);
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        ctx.write(msg, promise.addListener(channelFutureListener));
+    }
+
+    /**
+     * This method is called after receiving each {@link #channelRead} and should define if this read should trigger
+     * timeout cancelling. Default implementation returns {@code true}.
+     * @param ctx channel handler context of related {@link #channelRead} invocation
+     * @param msg message of related {@link #channelRead} invocation
+     * @return true if timeout should be cancelled by receiving this read, false otherwise
+     */
+    protected boolean timeoutShouldBeCancelled(ChannelHandlerContext ctx, Object msg) {
+        return true;
+    }
+
+    private void init(final ChannelHandlerContext ctx) {
+        removed = false;
+        if (channelFutureListener == null) {
+            channelFutureListener = new ChannelFutureListener() {
+                @Override
+                public void operationComplete(ChannelFuture future) throws Exception {
+                    if (future.isSuccess()) {
+                        initialize(ctx);
+                    }
+                }
+            };
+        }
+        if (readTimeoutTask == null) {
+            readTimeoutTask = new ReadTimeoutTask(ctx);
+        }
+    }
+
+    private void initialize(ChannelHandlerContext ctx) {
+        if (removed) {
+            return;
+        }
+
+        writeFinishTime = System.nanoTime();
+        readFinishTime = writeFinishTime - 1;
+        if (timeoutNanos > 0) {
+            timeout = ctx.executor().schedule(readTimeoutTask, timeoutNanos, TimeUnit.NANOSECONDS);
+        }
+    }
+
+    private void destroy() {
+        removed = true;
+        disable();
+    }
+
+    private void disable() {
+        if (timeout != null) {
+            timeout.cancel(false);
+            timeout = null;
+        }
+    }
+
+    /**
+     * Is called when a read timeout was detected.
+     */
+    private void readTimedOut(ChannelHandlerContext ctx) throws Exception {
+        if (!closed) {
+            ctx.fireExceptionCaught(ReadTimeoutException.INSTANCE);
+            if (closeOnTimeout) {
+                ctx.close();
+            }
+            closed = true;
```

Comment:
May be just use `ctx.channel.isOpen()` rather than having a local flag to this handler?